### ISSUE TITLE
 Unknown loss function error after load_model_hdf5()

### ISSUE THREAD

**Issue by juliavassey** [rstudio/keras#769]
Hi guys,

I would appreciate if you can take a look into the issue I'm having with loading the model trained using Inception V3. My training and validation data set consists of about 1,000 images of 5 categories.

The description of the issue is below. 

I trained the model using Inception V3 

``` r
# Compile the model (should be done *after* setting layers to non-trainable)
model %>%
  compile(optimizer =
            optimizer_adam(     
                           lr = 0.0001,                      
                           epsilon = 1e-08),    
          loss = loss_categorical_crossentropy,
          metrics = "accuracy")
```

Saved the model with this line of code:

``` r
model %>% save_model_hdf5("my_model_2.h5")
```
when trying to load the model from my working directory:

``` r
new_model <- load_model_hdf5("my_model_2.h5")
new_model %>% summary()
```

I'm getting the following error message:

```
Error in py_call_impl(callable, dots$args, dots$keywords) : 
  ValueError: Unknown loss function:loss_binary_crossentropy 
10.
stop(structure(list(message = "ValueError: Unknown loss function:loss_binary_crossentropy", 
    call = py_call_impl(callable, dots$args, dots$keywords), 
    cppstack = structure(list(file = "", line = -1L, stack = c("1   reticulate.so                       0x000000012af781eb _ZN4Rcpp9exceptionC2EPKcb + 267", 
    "2   reticulate.so                       0x000000012af78030 _ZN4Rcpp4stopERKNSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEE + 48",  ... 
9. deserialize_keras_object at generic_utils.py#193
8. deserialize at losses.py#183
7. get at losses.py#192
6. compile at training.py#466
5. _method_wrapper at base.py#474
4. load_model at saving.py#262
3. (structure(function (...) 
{
    dots <- py_resolve_dots(list(...))
    result <- py_call_impl(callable, dots$args, dots$keywords) ... 
2. do.call(keras$models$load_model, args) 
1. load_model_hdf5("my_model_2.h5") 
```

I'm using Anaconda, Python version 3.6.8. R packages are up-to-date.

The error message seems irrelevant, as I'm not using the loss_binary_crossentropy, but loss_categorical_crossentropy, which is a standard keras loss function.

Appreciate your response.

Julia

**Comment by skeydan** [comment]
Please provide the complete model definition.

**Comment by juliavassey** [comment]
base_model = applications.InceptionV3(weights='imagenet', include_top=False, input_shape=(img_width, img_height, 3))

It was built in R based on this Python script: 
https://github.com/paras42/Hello_World_Deep_Learning/blob/master/HelloWorldDeepLearning.ipynb

The only difference, my classifier is categorical, not binary (loss = categorical_crossenthropy)

**Comment by juliavassey** [comment]
**can't load the model and weights from my directory to R**

Hello, In addition to the above description, here are additional lines of codes to build the model:

base_model = applications.InceptionV3(weights='imagenet', include_top=False, input_shape=(img_width, img_height, 3))

freeze_weights(base_model)

model_top = base_model$output %>%
  layer_global_average_pooling_2d() %>%
  layer_dense(units = 128, activation = "relu") %>%
  layer_dropout(0.7) %>%
  layer_dense(units = 6, activation = "softmax")#output layer, sigmoid = probability between 0 and 1. This will get evaluated against the loss function 

model = keras_model(inputs = base_model$input, outputs = model_top)

length(model$layers)

# Manually freeze the original inception layers, just train the last 3 layers.
freeze_weights(model, 1, length(model$layers) - 3)
summary(model)

# Compile the model (should be done *after* setting layers to non-trainable)
model %>%
  compile(optimizer =
            optimizer_adam(#lr = 0.00001,
                           #lr = 0.0005,
                           lr = 0.00001,
                           # epsilon is a configuratoin setting for the adam optimizer.
                           epsilon = 1e-08),
          # Or can use string version: loss = "binary_crossentropy".
          loss = loss_categorical_crossentropy,
          metrics = "accuracy")

Appreciate any point in the right direction

**Comment by skeydan** [comment]
Sorry I don’t have an idea what could be going on here, apart from suspecting that perhaps something is wrong somehow with that file, or different versions of code might have gotten mixed up...

**Comment by juliavassey** [comment]
can't load the model and weights from my directory to R_UPDATE

To update this thread, i managed to load the model h5 from my directory after changing the loss function from the direct function name to a string, like so:
```
model %>%
  compile(optimizer =
            optimizer_adam(#lr = 0.00001,
                           #lr = 0.0005,
                           lr = 0.00001,
                           epsilon = 1e-08),
          #loss = loss_categorical_crossentropy,
          loss = "categorical_crossentropy",
          metrics = "accuracy")
```
Was wondering if you can consider fixing it so it works with either the string or the direct function name. 

In addition, I got this message below while loading the model, but I assume it has to do with the computation speed and doesn't imply any errors in the model, correct? 

```
new_model <- load_model_hdf5("my_model_7.h5")
```
2019-05-23 07:10:56.268990: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA

**Comment by skeydan** [comment]
> n addition, I got this message below while loading the model, but I assume it has to do with the computation speed and doesn't imply any errors in the model, correct?

Yeah exactly :-)

**Comment by juliavassey** [comment]
Is there a way for you to look at the second issue that I opened on Keras
yesterday?
https://github.com/rstudio/keras/issues/786

Thank you,

Sent from my iPhone

On May 24, 2019, at 8:11 AM, Sigrid Keydana <notifications@github.com>
wrote:

n addition, I got this message below while loading the model, but I assume
it has to do with the computation speed and doesn't imply any errors in the
model, correct?

Yeah exactly :-)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/rstudio/keras/issues/769?email_source=notifications&email_token=AHPG2QBOBRASZV3UAD4OKLDPXAATPA5CNFSM4HLNGHLKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWFVUUQ#issuecomment-495671890>,
or mute the thread
